h/views/api_auth.py

```diff
@@ -2,11 +2,111 @@
 
 from __future__ import unicode_literals
 
+from pyramid.httpexceptions import HTTPBadRequest, HTTPFound
+from pyramid.view import view_config, view_defaults
+
 from h.exceptions import OAuthTokenError
 from h.util.view import cors_json_view
 from h.util.datetime import utc_iso8601
 
 
+@view_defaults(route_name='oauth_authorize')
+class OAuthAuthorizeController(object):
+
+    def __init__(self, request):
+        self.request = request
+        self.oauth_svc = self.request.find_service(name='oauth')
+        self.user_svc = self.request.find_service(name='user')
+
+    @view_config(request_method='GET',
+                 renderer='h:templates/oauth/authorize.html.jinja2')
+    def get(self):
+        """
+        Check the user's authentication status and present the authorization
+        page.
+        """
+        self._check_params()
+
+        if self.request.authenticated_userid is None:
+            return HTTPFound(self.request.route_url('login', _query={
+                              'next': self.request.url}))
+
+        params = self.request.params
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        return {'username': user.username,
+                'client_name': 'Hypothesis',
+                'client_id': params['client_id'],
+                'response_type': params['response_type'],
+                'response_mode': params['response_mode'],
+                'origin': params.get('origin'),
```

Comment:
We might not need this any more. OAuth clients have redirect URIs or origins associated with them when they are registered. _If_ the service allows a client to have multiple redirect URIs/origins associated with it then this tells the service which specific URI or origin to send the response back to. What we're currently planning is that initially we'll only allow one redirect URI or origin for each auth client, in which case this param is redundant because we can determine it from the `client_id`.